File: src/main/java/com/matag/game/player/Battlefield.java
package com.matag.game.player;

import java.util.List;
import java.util.stream.Collectors;

import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import com.matag.game.cardinstance.CardInstance;
import com.matag.game.cardinstance.CardInstanceSearch;
import com.matag.game.cardinstance.CardListComponent;

@Component
@Scope("prototype")
public class Battlefield extends CardListComponent {

  public void removeAttacking() {
    new CardInstanceSearch(cards).attacking().getCards()
      .forEach(cardInstance -> cardInstance.getModifiers().setAttacking(false));
  }

  public void removeBlocking() {
    new CardInstanceSearch(cards).blocking().getCards()
      .forEach(cardInstance -> cardInstance.getModifiers().unsetBlockingCardId());
  }

  public List<CardInstance> getAttackingCreatures() {
    return new CardInstanceSearch(cards).attacking().getCards();
  }

  public List<CardInstance> getBlockingCreatures() {
    return new CardInstanceSearch(cards).blocking().getCards();
  }

  public List<CardInstance> getBlockingCreaturesFor(int attackingCardId) {
    return getBlockingCreatures().stream()
      .filter(cardInstance -> cardInstance.getModifiers().getBlockingCardId() == attackingCardId)
      .collect(Collectors.toList());
  }
}

File: src/main/kotlin/com/matag/game/player/Battlefield.kt
package com.matag.game.player

import com.matag.game.cardinstance.CardInstance
import com.matag.game.cardinstance.CardInstanceSearch
import com.matag.game.cardinstance.CardListComponent
import java.util.function.Consumer

class Battlefield : CardListComponent() {
    fun removeAttacking() {
        CardInstanceSearch(cards).attacking().cards
            .forEach(Consumer { cardInstance: CardInstance? -> cardInstance!!.modifiers.isAttacking = false })
    }

    fun removeBlocking() {
        CardInstanceSearch(cards).blocking().cards
            .forEach(Consumer { cardInstance: CardInstance? -> cardInstance!!.modifiers.unsetBlockingCardId() })
    }

    val attackingCreatures: List<CardInstance>
        get() = CardInstanceSearch(cards).attacking().cards

    val blockingCreatures: List<CardInstance>
        get() = CardInstanceSearch(cards).blocking().cards

    fun getBlockingCreaturesFor(attackingCardId: Int): List<CardInstance> {
        return this.blockingCreatures
            .filter { it.modifiers.blockingCardId == attackingCardId }
    }
}
